packages/story-editor/src/components/library/panes/media/common/insertionMenu.js

```diff
@@ -0,0 +1,200 @@
+ */
+function InsertionMenu({ resource, display, onInsert, width, index }) {
+  const insertButtonRef = useRef();
+  const [isMenuOpen, setIsMenuOpen] = useState(false);
+  const onMenuOpen = useCallback((e) => {
+    e.stopPropagation();
+    setIsMenuOpen(true);
+  }, []);
+  const onMenuCancelled = useCallback(() => setIsMenuOpen(false), []);
+
+  const { currentBackgroundId, combineElements } = useStory((state) => ({
+    currentBackgroundId: state.state.currentPage?.elements?.[0]?.id,
+    combineElements: state.actions.combineElements,
+  }));
+
+  const { type, poster } = resource;
+  const insertLabel = ['image', 'gif'].includes(type)
+    ? __('Insert image', 'web-stories')
+    : __('Insert video', 'web-stories');
+  const options = [
+    {
+      group: [
+        { label: insertLabel, value: 'insert' },
+        {
+          label: __('Add as background', 'web-stories'),
+          value: 'addBackground',
+        },
+      ],
+    },
+  ];
+
+  const handleCurrentValue = (evt, value) => {
+    const thumbnailUrl = poster
+      ? poster
+      : getSmallestUrlForWidth(width, resource);
+    const newElement = getElementProperties(resource.type, {
+      resource,
+    });
+    switch (value) {
+      case 'insert':
+        onInsert(resource, thumbnailUrl);
+        break;
+      case 'addBackground':
+        combineElements({
+          firstElement: newElement,
+          secondId: currentBackgroundId,
+        });
+        break;
+      default:
+        break;
+    }
+  };
+
+  useNestedRovingTabIndex({ ref: insertButtonRef }, [], 3);
+
+  const listId = useMemo(() => `list-${uuidv4()}`, []);
+  const buttonId = useMemo(() => `button-${uuidv4()}`, []);
+
+  // Keep icon and menu displayed if menu is open (even if user's mouse leaves the area).
+  return (
+    <MenuContainer>
+      <AddButton
+        ref={insertButtonRef}
+        onClick={onMenuOpen}
+        aria-label={__('Open insertion menu', 'web-stories')}
+        aria-pressed={isMenuOpen}
+        aria-haspopup
+        aria-expanded={isMenuOpen}
+        aria-owns={isMenuOpen ? listId : null}
+        id={buttonId}
+        display={display}
+        tabIndex={index === 0 ? 0 : -1}
+      >
+        <IconContainer>
+          <Icons.PlusFilled />
+        </IconContainer>
+      </AddButton>
+      {(display || isMenuOpen) && (
+        <Popup
+          anchor={insertButtonRef}
+          placement={PLACEMENT.BOTTOM_START}
+          isOpen={isMenuOpen}
+        >
+          <DropDownContainer>
+            <Menu
+              parentId={buttonId}
+              listId={listId}
+              onMenuItemClick={handleCurrentValue}
+              options={options}
+              onDismissMenu={onMenuCancelled}
+              hasMenuRole
+              menuStylesOverride={menuStylesOverride}
+            />
+          </DropDownContainer>
+        </Popup>
+      )}
+    </MenuContainer>
+  );
+}
+
+InsertionMenu.propTypes = {
+  resource: PropTypes.object.isRequired,
```

Comment:
could you use this `ResourcePropTypes` (found it referenced in story-editor types) so avoid the blanket `object` shape?